### PR TITLE
Define a method `ConfigCursor#atPath`

### PR DESCRIPTION
**Pull request by missingfaktor**
Similar to `ConfigCursor#atKey` but for deeper nested paths. We have this in our project's utils, but were wondering if this could be put in PureConfig itself.

### CODE REVIEW

**Comment by ruippeixotog** [comment]
That's a great idea, thanks for the contribution!

I have only one comment: I think this can be part of `ConfigCursor`, and not of `ConfigObjectCursor`. `atKey` is kind of a primitive operation for cursors - since we forced the user to cast the cursor to an object cursor first, that operation does a very specific thing, and without this operation we can't fully navigate through the config. On the other hand, `atPath` already has to do "casts" internally; since the operation can already fail with `WrongType`, I'd say it makes sense to make it even easier to use and allow it on generic cursors.

Incidentally, that would also greatly simplify your code, since you wouldn't have to handle the last key differently anymore.

**Comment by missingfaktor** [comment]
Okay, let me submit a change in a separate commit. We can squash later if you approve of this change.

File: core/src/main/scala/pureconfig/ConfigCursor.scala
Comment (ruippeixotog) [inline core/src/main/scala/pureconfig/ConfigCursor.scala]: Now that the method is in `ConfigCursor`, you don't need to extract and treat `lastKey` in a separate way. You can simplify this expression by folding over `key +: moreKeys` with an initial value of `this`.

**Review by ruippeixotog — CHANGES_REQUESTED** [review]
Since `Either` in Scala 2.11 and below isn't right-biased, you have to call `.right` before each `.flatMap` on an `Either`.

**Comment by missingfaktor** [comment]
Thanks, @ruippeixotog. I have added an improved version. A few notes about it:
- It can navigate not only through objects but arrays too; something I missed earlier.
- The new definition is recursive. A simple reasoning shows it will terminate, but I am unable to convince the Scala compiler of it. If you could suggest changes that would make it tail-recursive, that would be great.
- I am using `implicitConversions` here. Sadly Scala's `implicit class` mechanism does not cater to sum types, even when newly defined, so I had to use this special language feature here. Hope that's alright.

File: core/src/main/scala/pureconfig/Path.scala
Comment (ruippeixotog) [inline core/src/main/scala/pureconfig/Path.scala]: Is there any advantage of having this class instead of using always a `Seq[Path.Segment]`? You could have a single `atPath` method if you were to remove this.

File: core/src/main/scala/pureconfig/syntax/package.scala
Comment (ruippeixotog) [inline core/src/main/scala/pureconfig/syntax/package.scala]: These conversions should be better inside the companion object of `Path` or `Path.Segment`. They do not need to be imported explicitly if they're there.

File: core/src/main/scala/pureconfig/ConfigCursor.scala
Comment (ruippeixotog) [inline core/src/main/scala/pureconfig/ConfigCursor.scala]: In the general case, monads such as `Either` are not tail-recursive, so you can't easily make this stack safe. Can you revert this to the old `foldLeft` formulation?

**Review by ruippeixotog — CHANGES_REQUESTED** [review]

File: core/src/main/scala/pureconfig/Path.scala
Comment (missingfaktor) [reply core/src/main/scala/pureconfig/Path.scala]: The only advantage is a separation between a more fundamental method (one using `Path`) and one defined in a way to make it syntactically more convenient for direct use (one using `Path.Segment*`). The former is more suitable for composition. But we can always reintroduce it when needed. For now I will remove it.

File: core/src/main/scala/pureconfig/syntax/package.scala
Comment (missingfaktor) [reply core/src/main/scala/pureconfig/syntax/package.scala]: Okay, will do.

File: core/src/main/scala/pureconfig/ConfigCursor.scala
Comment (missingfaktor) [reply core/src/main/scala/pureconfig/ConfigCursor.scala]: Okay, will do.

**Comment by missingfaktor** [comment]
@ruippeixotog I have made the requested changes. Please re-review!

**Review by ruippeixotog — APPROVED** [review]
LGTM, thanks again for the contribution!

**Comment by missingfaktor** [comment]
Thanks, @ruippeixotog! Do you need me to add documentation, CHANGELOG entries etc, or is this ready for merge as it is?

**Comment by ruippeixotog** [comment]
I don't think so, don't worry about it :) We typically wait for the approvals of two authors before merging non-trivial changes, so let's wait until someone else reviews this.

**Comment by missingfaktor** [comment]
Alright, no problem. Thanks again. :)

File: core/src/test/scala/pureconfig/ConfigCursorSuite.scala
Comment (jcazevedo) [inline core/src/test/scala/pureconfig/ConfigCursorSuite.scala]: Can you please move this to the behavior of `ConfigCursor` (and also use `cursor` instead of `objCursor` to create the test configurations)? The `at` method isn't specific of `ConfigObjectCursor`.

File: core/src/main/scala/pureconfig/ConfigCursor.scala
Comment (jcazevedo) [inline core/src/main/scala/pureconfig/ConfigCursor.scala]: Can you please also add Scaladoc for these methods?

File: core/src/main/scala/pureconfig/ConfigCursor.scala
Comment (jcazevedo) [inline core/src/main/scala/pureconfig/ConfigCursor.scala]: Even though the compiler always chooses the most specific method to call, having this overloaded method in the public API of `ConfigCursor` might be a bit confusing to read. What do you think of making this definition private (or just inlining it in the other `at` definition)?

**Review by jcazevedo — CHANGES_REQUESTED** [review]
Thanks for your contribution! Any reason why you prefer `at` instead of `atPath`? I think `atPath` describes the functionality more clear and is more consistent with the `atKey` and `atIndex` methods of `ConfigObjectCursor` and `ConfigListCursor`.

File: core/src/test/scala/pureconfig/ConfigCursorSuite.scala
Comment (missingfaktor) [reply core/src/test/scala/pureconfig/ConfigCursorSuite.scala]: Good point.

File: core/src/main/scala/pureconfig/ConfigCursor.scala
Comment (missingfaktor) [reply core/src/main/scala/pureconfig/ConfigCursor.scala]: Sure thing.

File: core/src/main/scala/pureconfig/ConfigCursor.scala
Comment (missingfaktor) [reply core/src/main/scala/pureconfig/ConfigCursor.scala]: * I named it `at` because it is a more general version of `atKey` and `atIndex`. `*Path` suffix didn't seem to add anything of value. But if you think it makes it more readable, happy to change the name back.
* I would keep the two methods separate because each does one specific thing, and I think that makes them easier to understand. I can make the non-varargs one private as the varargs one subsumes it for the users.

**Comment by missingfaktor** [comment]
@jcazevedo All review points addressed.

**Comment by missingfaktor** [comment]
Hello, @jcazevedo and @ruippeixotog. Do you need me to make any more changes?

**Review by jcazevedo — APPROVED** [review]

**Comment by jcazevedo** [comment]
> Hello, @jcazevedo and @ruippeixotog. Do you need me to make any more changes?

No. It's good. Thanks again for your contribution, and sorry for the delay in reviewing this!

**Comment by missingfaktor** [comment]
Thanks, @jcazevedo! :-)

**Pull request merged**
